File: src/apps/unittests/tests/tenvvars.cc
/*
 * Copyright (C) 2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#if defined(__kachel__)
#   define _GNU_SOURCE
#   include <stdlib.h>
#endif

#include <base/Common.h>

#include <m3/tiles/ChildActivity.h>
#include <m3/EnvVars.h>
#include <m3/Test.h>

#include "../unittests.h"

using namespace m3;

static void basics() {
    WVASSERTSTREQ(EnvVars::get("FOO"), nullptr);
    EnvVars::set("TEST", "value");
    WVASSERTSTREQ(EnvVars::get("TEST"), "value");

    WVASSERTEQ(EnvVars::count(), 1u);
    auto vars = EnvVars::vars();
    WVASSERTSTREQ(vars[0], "TEST=value");
    WVASSERTSTREQ(vars[1], nullptr);

    EnvVars::remove("ABC");
    WVASSERTEQ(EnvVars::count(), 1u);
    EnvVars::remove("TEST");
    WVASSERTEQ(EnvVars::count(), 0u);
    WVASSERTSTREQ(EnvVars::get("FOO"), nullptr);
}

static void multi() {
    EnvVars::set("V1", "val1");
#if defined(__kachel__)
    setenv("V2", "val2", 1);
#else
    EnvVars::set("V2", "val2");
#endif
    EnvVars::set("V2", "val3");
    EnvVars::set("V21", "val=with=eq");
    WVASSERTEQ(EnvVars::count(), 3u);

    {
        auto vars = EnvVars::vars();
        WVASSERTSTREQ(vars[0], "V1=val1");
        WVASSERTSTREQ(vars[1], "V2=val3");
        WVASSERTSTREQ(vars[2], "V21=val=with=eq");
        WVASSERTSTREQ(vars[3], nullptr);
    }

    EnvVars::remove("V1");
    WVASSERTEQ(EnvVars::count(), 2u);
    {
        auto vars = EnvVars::vars();
        WVASSERTSTREQ(vars[0], "V2=val3");
        WVASSERTSTREQ(vars[1], "V21=val=with=eq");
        WVASSERTSTREQ(vars[2], nullptr);
    }

#if defined(__kachel__)
    unsetenv("V21");
#else
    EnvVars::remove("V21");
#endif
    WVASSERTEQ(EnvVars::count(), 1u);
    {
        auto vars = EnvVars::vars();
        WVASSERTSTREQ(vars[0], "V2=val3");
        WVASSERTSTREQ(vars[1], nullptr);
    }

    EnvVars::remove("V2");
    WVASSERTEQ(EnvVars::count(), 0u);
    {
        auto vars = EnvVars::vars();
        WVASSERTSTREQ(vars[0], nullptr);
    }
}

static void to_child() {
    EnvVars::set("V1", "val1");
    EnvVars::set("V2", "val2");
    EnvVars::set("V3", "val3");

    ChildActivity act(Tile::get("clone|own"), "child");

    act.run([] {
        auto vars = EnvVars::vars();
        WVASSERTEQ(EnvVars::count(), 3u);
        WVASSERTSTREQ(vars[0], "V1=val1");
        WVASSERTSTREQ(vars[1], "V2=val2");
        WVASSERTSTREQ(vars[2], "V3=val3");
        WVASSERTSTREQ(vars[3], nullptr);
        EnvVars::remove("V2");
        WVASSERTEQ(EnvVars::count(), 2u);
        return 0;
    });

    WVASSERTEQ(act.wait(), 0);

    EnvVars::remove("V2");
    EnvVars::remove("V3");
    EnvVars::remove("V1");
    WVASSERTEQ(EnvVars::count(), 0u);
}

void tenvvars() {
    RUN_TEST(basics);
    RUN_TEST(multi);
    RUN_TEST(to_child);
}

File: src/apps/rustunittests/src/tenvvars.rs
/*
 * Copyright (C) 2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use m3::col::ToString;
use m3::env;
use m3::errors::Code;
use m3::test::{DefaultWvTester, WvTester};
use m3::tiles::{ActivityArgs, ChildActivity, RunningActivity, Tile};
use m3::{wv_assert_eq, wv_assert_ok, wv_run_test};

pub fn run(t: &mut dyn WvTester) {
    wv_run_test!(t, basics);
    wv_run_test!(t, multi);
    wv_run_test!(t, to_child);
}

fn basics(t: &mut dyn WvTester) {
    wv_assert_eq!(t, env::var("FOO"), None);
    env::set_var("TEST", "value");
    wv_assert_eq!(t, env::var("TEST"), Some("value".to_string()));

    wv_assert_eq!(t, env::vars().len(), 1);
    let vars = env::vars();
    let mut it = vars.iter();
    wv_assert_eq!(
        t,
        it.next(),
        Some(&("TEST".to_string(), "value".to_string()))
    );
    wv_assert_eq!(t, it.next(), None);

    env::remove_var("ABC");
    wv_assert_eq!(t, env::vars().len(), 1);
    env::remove_var("TEST");
    wv_assert_eq!(t, env::vars().len(), 0);
    wv_assert_eq!(t, env::var("TEST"), None);
}

fn multi(t: &mut dyn WvTester) {
    env::set_var("V1", "val1");
    env::set_var("V2", "val2");
    env::set_var("V2", "val3");
    env::set_var("V21", "val=with=eq");
    wv_assert_eq!(t, env::vars().len(), 3);

    let vars = env::vars();
    let mut it = vars.iter();
    wv_assert_eq!(t, it.next(), Some(&("V1".to_string(), "val1".to_string())));
    wv_assert_eq!(t, it.next(), Some(&("V2".to_string(), "val3".to_string())));
    wv_assert_eq!(
        t,
        it.next(),
        Some(&("V21".to_string(), "val=with=eq".to_string()))
    );
    wv_assert_eq!(t, it.next(), None);

    env::remove_var("V2");
    wv_assert_eq!(t, env::vars().len(), 2);
    env::remove_var("V21");
    wv_assert_eq!(t, env::vars().len(), 1);
    env::remove_var("V1");
    wv_assert_eq!(t, env::vars().len(), 0);
}

fn to_child(t: &mut dyn WvTester) {
    env::set_var("V1", "val1");
    env::set_var("V2", "val2");
    env::set_var("V3", "val3");

    let act = wv_assert_ok!(ChildActivity::new_with(
        wv_assert_ok!(Tile::get("compat|own")),
        ActivityArgs::new("child")
    ));

    let run = wv_assert_ok!(act.run(|| {
        let mut t = DefaultWvTester::default();
        wv_assert_eq!(t, env::vars().len(), 3);
        let vars = env::vars();
        let mut it = vars.iter();
        wv_assert_eq!(t, it.next(), Some(&("V1".to_string(), "val1".to_string())));
        wv_assert_eq!(t, it.next(), Some(&("V2".to_string(), "val2".to_string())));
        wv_assert_eq!(t, it.next(), Some(&("V3".to_string(), "val3".to_string())));
        wv_assert_eq!(t, it.next(), None);

        env::remove_var("V2");
        wv_assert_eq!(t, env::vars().len(), 2);
        Ok(())
    }));

    wv_assert_eq!(t, run.wait(), Ok(Code::Success));

    env::remove_var("V3");
    env::remove_var("V2");
    env::remove_var("V1");
    wv_assert_eq!(t, env::vars().len(), 0);
}
